gtfu/src/main/java/gtfu/tools/TripListGenerator.java

```diff
@@ -177,6 +192,26 @@ private static Map<String, String> createCustomDirectionMap() {
         return directionMap;
     }
 
+    private static Map<String, List<String>> createServiceIDMap(int exception, CalendarDateCollection calendarDates){
+        Map<String, List<String>> dateMap = new HashMap();
+        CalendarDate currentCalDate = null;
+
+        for(int i = 0; i < calendarDates.getSize(); i++){
+            currentCalDate = calendarDates.get(i);
+            if(currentCalDate.getExceptionType() == exception){
+                if(dateMap.get(currentCalDate.getServiceID()) == null){
```

Comment:
This becomes less redundant when written as:
```
Item item = cache.get(key);

if (item == null) {
    item = new Item();
    cache.put(key, item);
} 

doStuffWith(item);
```

This is a pattern that happens pretty frequently, which is why I wrote it in a more generalized way.